Js tradepart2

jacobstringer opened the pull request:
This PR includes an earlier PR and lots of other stuff.

- Trade rewritten, now using recursion, pruning and memoisation to come to an optimal and complete solution every time, includes more information such as a message about the card play, and includes all the functionality that was catered for by check_valid_move previously
- Process cards/play card rewritten but maintaining backward functionality for the AI which still uses the old version (while Sam is working on that section)
- Play_card API now returns information about the move to the front end (Marthijn has created a diagram to explain how this is to be processed on the front end)
- Tests have been updated due to the different interfaces being used for process_card and trading now
- Fixes a couple of bugs

Front end says it's an invalid move each move, due to the new information given back from play_card. It works fine though and will just need some adjustments to handle the additional information.

jacobstringer committed:
Refactors trade logic to a complete solution.

Not yet done:
- Piping through to main game

jacobstringer committed:
Refactors trade logic to a complete solution.

Not yet done:
- Piping through to main game

jacobstringer committed:
Refactors trade logic to a complete solution.

Not yet done:
- Piping through to main game

jacobstringer committed:
Merge branch 'JS-traderefactor'

jacobstringer committed:
update

jacobstringer committed:
Creates new 'play_card_with_trade' which uses the new trade/move checking function for player moves while still maintaining backward compatibility with the AI section (which still uses the old play_card and check_move functions) until a refactor there can be completed.

Pipes play_card API through to all the move information needed for UX design

Moves trade into its own file

jacobstringer committed:
Updates the back end tests to work with the new process_card returns.

Sets the default players in a single player game to 3.

jacobstringer committed:
Merge branch 'master' into JS-tradepart2

mmdonaldson commented:
@Sam-Irv can you please review

mmdonaldson commented on application/app.py:
```diff
@@ -268,13 +273,7 @@ def play_card():
-            status="Invalid move",
-            game=print_json(player)
-        )
+    return jsonify(process_card(card, player, discarded, for_wonder, from_discard_pile=from_discard_pile, process_with_trade=trade))
```
What if the move is invalid? It looks like you are removing the else statment

jacobstringer replied:
Check out the data flow,  you'll see both valid and invalid moves are accounted for.

Marthijn-B replied:
Check out ClientServer_Playcard_Dataflow.PDF in https://drive.google.com/drive/folders/0BxaJR_flclorQ3RYaUs4UmREamM

Marthijn-B committed:
Change ai_players string to int

Code was breaking because of it
Wasn't expecting string...
User selection of number AI players are working

Sam-Irv reviewed (approved):
Looks fine to me.

Should the front end be updated before pushing this to master?

Marthijn-B committed:
Updated Actions to access new Data object for Play_Card

Marthijn-B committed:
Updated RECEIVE_PLAY_CARD for new trade data

to send to browser for rendering

jacobstringer commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add feedback about turns.](https://trello.com/c/erW19quK/51-add-feedback-about-turns)

jacobstringer commented:
![](https://github.trello.services/images/mini-trello-icon.png) [UI: Allow the user to decide how many players to play with on single player mode.](https://trello.com/c/PuMnIho5/74-ui-allow-the-user-to-decide-how-many-players-to-play-with-on-single-player-mode)

Marthijn-B committed:
Dialog for trading

Call playcard with trade=false.
Dialog to suggest payment
if user continue, playcard api with trade=true.
This is the same to play wonder cards and buying resources

Marthijn-B committed:
Re-upload

Incase I have changed something in the last few hours

Marthijn-B committed:
Re-upload for slight changes

introducing trade reference

Marthijn-B committed:
Removed commented code

Marthijn-B commented:
Trading Dialog on the front End is complete. I have not merged this PR yet. 
Could someone look over it before it is merged? 
(I know it has been approved yet for the back end  part already)

mmdonaldson commented on static/src/components/GameScreen.js:
```diff
@@ -224,17 +332,22 @@ export class GameScreen extends Component {
-                if (nextWonderLevel > game.player.max_wonder) {
-                  canPlayWonder = false;
-                }
+        if (started && game.game.age) {
```
You've overwritten Jacobs change on the game check. If you access game.game.age and game is null then this will through a null reference exception

Marthijn-B replied:
I hear ya... Changed

Marthijn-B committed:
Changed line 335 back to Started && game

Marthijn-B committed:
Display Successful card play messages

I added this as I haven't utilized the card play success.
Adds messages between 'Cards in Hand' and 'Player Display' section
(Last change for this PR- I promise)

jacobstringer committed:
Updated prereq message to include the cardname played

mmdonaldson commented on static/src/components/GameScreen.js:
```diff
@@ -332,7 +332,7 @@ export class GameScreen extends Component {
             this.playersLogged();
         }
-        if (started && game.game.age) {
+        if (started && game) {
```
I think this needs to be:
```
if (started && game) {
  if (game.game.age) {
    //
  }
}
```

jacobstringer replied:
Hmmm it must've been messed with by all the commits and merging that have happened. Will sort now.

jacobstringer committed:
Update GameScreen.js
